Add game features, update graphics

bquan0 opened the pull request:
Closes #6, #8, #9, #11.
## New Features
### Main Menu Screen
- Has background animations of a camera panning through Level 1.
- Buttons to quit the game and go to the level select screen.

### Level Select Screen
- Has a scroll menu to navigate through the various levels.
- When a level is hovered over, a thumbnail of that level will be displayed to the right of the scroll menu.

### Loading Screen
- Used when loading a level (since loading the data from a `.csv` file into an array takes a relatively large amount of time)
- Has a progress bar (whose function isn't really shown because loading the `.csv` file takes up most of the time and it happens first)

### Pause Screen
- Displays the instructions and objectives of the current level.
- The Pause Screen is a scene of its own that can be reused in every level. 
- Each level requires its own Instructions and Objectives scenes.
- Buttons to quit the game and go to the level select screen

### Objectives
For Level 1, there are two objectives: 
- "Reach the End Zone" detection is achieved using an Area node to detect when the Player has entered the End Zone.
- "See the Radioactive Cylinder" detection is achieved using raycasting and checking if there's no walls between the Player and the Cylinder.
Completing an objective will cause a notification to be displayed and will mark the objective as completed in the Pause Screen

### End Screen
- Shows the total amount of radiation received by the Player (currently still actually the Geiger Counter) and the objectives 
- Buttons to go back to Level Select, Quit, or Continue (go to next level)

### Models
- Updated the Geiger Counter Model.
- Added PBR textures and collisions to the TwoRoom model.
- Expanded the layout of the TwoRoom model to include a hallway, starting zone, and ending zone.

## Code Review
General thoughts:
- Changed `Main` scene's name to `Level1` since the new default scene is the `MainMenu` scene.
- Many scripts now have `onready var` for the paths to certain nodes. This is to increase readability and also performance because we only have to find the node once at the start.
- Deleted `GeigerViewport.gd` because I realized that the tutorial I followed for 3D Labels was outdated and `Label3D` exists starting in Godot 3.5 (the version I'm using)

### `Global.gd`
- This is a [Godot Singleton](https://docs.godotengine.org/en/3.5/tutorials/scripting/singletons_autoload.html) which acts like a library of functions and variables that can be accessed by any script. Ex: `Global.go_to_scene("...")`
- `load_scene()` loads a scene in the background while displaying the loading screen. Used for loading in scenes with radiation maps that need data from `.csv` files (since they load slowly)
- `go_to_scene()` switches to a scene directly without showing the loading screen. The deferred method is there just to make sure that nothing is running when we try to switch the current scene.

### `MainMenu.gd`
The reason why we don't call `get_tree().quit()` in the `_on_QuitButton_pressed()` function and instead send a notification so that the game quits in the `_notification()` function is because in the future, we might want to do things before we quit, such as saving Player data and whatnot.

### `ObjNotif.gd`
This uses an array as a queue to store all the objectives because it's possible for the player to complete multiple objectives at the same time. Without the queue, only the first objective completed will send a notification, and the other objective notifications won't be shown. This could be updated in the future to use a `ScrollContainer` so that multiple objectives can be shown at once.

### `Player.gd`
Removed the `GC_click_sound` variable because it wasn't used in this script, so its existence seemed arbitrary. Don't worry, it's been added to `Level1.gd` since that's where it was actually being used. 

### `text_theme.tres`
This is a `Theme` that I'm using for the buttons in `MainMenu` and `LevelSelect`. Originally, this was so that I could reuse the same theme in different scenes, but then I realized that I could just copy and paste themes across different scenes, so I didn't save any more of my themes as `.tres`. This could be removed if necessary, it would just take a little bit of effort to remake the theme and not save it.


## Thoughts, Questions
### Modularity
Previously, there was the idea of being able to plug in `.csv` files into the game in order to simulate different radiation situations. After texturing the TwoRoom model, I do not think this is a feasible idea without some advanced infrastructure to generate the associated models, textures, and collisions in Godot. This is because the `.csv` file will get you the radiation map, but it doesn't have any information about the collisions (which must be done in Godot), model, or textures.

### Quit, LevelSelect Buttons
Currently, the Quit and Level Select Buttons have similar functionalities wherever they are. Should they be made into their own scenes for more modularity and not having to repeat these functions in other scenes? Currently, they are being used in the `MainMenu`, `PauseScreen`, and `LvlCompleteScreen` scenes, and it's unlikely they will be used in another scene. Also, making them scenes means 2 extra scripts and scenes. 

## Final Remarks
If anyone has any ideas about what I should add to the game, please add a draft or issue in the [GitHub Project](https://github.com/orgs/cnerg/projects/9/views/1).

bquan0 committed:
make the game look prettier

bquan0 committed:
remove Wall.tscn and simplify PauseScreen hierarchy

gonuke commented on .gitignore:
```diff
@@ -20,4 +20,6 @@ mono_crash.*.json
 *.obj
-*.ogg
\ No newline at end of file
+*.ogg
+*.png
+*.jpg
```
I think we discussed this before, but don't we want include these artifacts to ensure the reproducibility of the project?

bquan0 replied:
They are not included for the same reason as the `.csv`, `.ogg`, and `.obj` files, which is that they can be large and take up too much space in the repo.

gonuke replied:
Are they all large?  I think we need to find some solution to this.  Can you post some information about file sizes?

bquan0 replied:
`.csv`
- `points.csv` is 222KB
- `tank10.csv` is 223MB (large)
- `tworoom10msht.csv` is 28MB

`.obj`
- `GeigerCounter.obj` is 3MB
- `TwoRoomModel.obj` is 43KB

`.ogg`
- `geigercounter.ogg` is 6 KB

`.jpg`
- In total, the `textures` folder is 18.5MB
- Each `.jpg` ranges from 0.4MB to 2MB, and there are 4 `.jpg` used for each PBR texture. There are currently 5 PBR textures in game. 

The `fonts` folder is 451KB. 

Another thing to consider is that the `.git` folder will become very large if there are ever any changes made to these large binary files.

gonuke replied:
Yeah... a little larger than we'd like and I don't want to get into git large file stuff...

That said, we would like to have an easy way for someone else to replicate this. We can host these files somewhere and have a step to download and unpack them into the right place.

We may also want to think about how we separate them out.  For example, I could imagine building a library of textures (jpg) and sounds (ogg), but the geometry (obj) and rad data (csv) are unique to a particular case.

gonuke commented on scripts/Global.gd:
```diff
@@ -0,0 +1,82 @@
+		if status == OK:
+			# update progress bar according to amount of data loaded
+			var progress_bar = loading_scene_instance.get_node("./CanvasLayer2/ProgressBar")
+			progress_bar.value = float(loader.get_stage()) / loader.get_stage_count() * 100
```
Can you query and store `loader.get_stage_count()` outside the loop, and even store the inverse * 100 to avoid the frequent operations?

gonuke replied:
(maybe this loop isn't run very often???)

bquan0 replied:
`status == OK:` is a situation that only happens when the loader is done loading a chunk of the scene, so it doesn't happen that often. I will still make those changes though.

gonuke commented on scripts/PauseScreen.gd:
```diff
@@ -0,0 +1,34 @@
+	if event.is_action_pressed("esc"):
+		# toggle whether the game is paused or not
+		show_pause = !show_pause
+		visible = show_pause
```
Out of curiosity -  is `visible` a member variable of some kind that affects the state?

bquan0 replied:
`visible` is a Godot variable for Nodes that controls whether they are shown or hidden in the game viewport

gonuke commented on scripts/Level1.gd:
```diff
@@ -0,0 +1,145 @@
+	# check if there's nothing between Camera and Cylinder
+	if obj_between and obj_between.collider == Cylinder:
+		# check if the Cylinder is in the Camera's FOV (using dot product and angles)
+		var cam_vector = -1 * PlayerCamera.get_global_transform().basis.z
+		var cam_to_cylnd_vector = Cylinder.global_transform.origin - PlayerCamera.global_transform.origin
+		var dot_product = cam_vector.dot(cam_to_cylnd_vector)
+		var angle = acos(dot_product / (cam_vector.length() * cam_to_cylnd_vector.length()))
+		var angle_deg = rad2deg(angle)
+		# player saw the Cylinder if the Cylinder is in Camera's FOV
+		if angle_deg <= half_camera_fov and angle_deg >= (-1) * half_camera_fov:
+			# make sure the objective cannot be completed more than once
+			saw_cylinder = true
+			# check off the objective
+			SeeCylinderObj.pressed = true
+			# show a notification for the objective that was completed
+			$ObjNotif.add_notification(SeeCylinderObj.text)
```
It would seem useful to put this in a function so that you can easily apply it to test if another object is "seen".

gonuke commented on scripts/Level1.gd:
```diff
@@ -0,0 +1,145 @@
+		
+	# show LvlCompleteScreen if the player presses "f" in the EndZone
+	if CompLvlText.visible and Input.is_action_pressed("interact") and !lvl_complete:
+		# prevent this from happening more than once
+		lvl_complete = true
+		# remove PauseScreen to prevent player from pressing "esc" to open it
+		$PauseScreen.queue_free()
+		# prevent TabScreen from being opened
+		$TabScreen.lvl_complete = true
+		# reparent Objectives node (getchild, removechild, addchild)
+		var objectives = $InstObj1/VBoxContainer/Objectives1
+		$InstObj1/VBoxContainer.remove_child(objectives)
+		LvlCompleteScreen.add_child(objectives)
+		# move objectives to a better position
+		objectives.rect_position = RadiationReceived.rect_position
+		objectives.rect_position.y += RadiationReceived.rect_size.y + 20
+		
+		# pause the game
+		get_tree().paused = true
+		# let player move mouse (in case he is fully zoomed in)
+		Input.set_mouse_mode(Input.MOUSE_MODE_VISIBLE)
+		
+		# show the screen
+		LvlCompleteScreen.show()
```
Make this a function for readability?

gonuke reviewed (changes requested):
This looks pretty good, mostly very minor points, and some may not make sense - I'm guessing some things about how Godot works...

bquan0 committed:
add functions for readability and reusability

bquan0 committed:
fix bugs with LvlCompleteScreen

bquan0 committed:
add README.md, geigercounter.ogg, HowToPlayScreen

bquan0 committed:
add resources to README.md

bquan0 commented:
I made the following changes:

- Added a `README.md` to explain how to setup the project and some extra resources for people who are trying to understand the code in the repo. 
- Cleaned up the folder structure and posted the extra resources not included in this repo to a [folder in my UW Google drive](https://drive.google.com/drive/u/2/folders/10QSwU5U0PyNMef77HWv0K5GObslwSkSL). I decided to put `geiger_counter.ogg` in the repo since it's only 6 KB and it's the only sound effect in the game. 
- Added a How To Play screen to show players the controls of the game.
